stdlib/builtins.pyi

```diff
@@ -960,7 +960,7 @@ class tuple(Sequence[_T_co], Generic[_T_co]):
     @overload
     def __getitem__(self, __key: SupportsIndex) -> _T_co: ...
     @overload
-    def __getitem__(self, __key: slice) -> tuple[_T_co, ...]: ...
+    def __getitem__(self, __key: slice) -> Self: ...
```

Comment:
That would imply that this should hold true, which I don't think it's accurate:

```python
def func(t: tuple[int, int, int]):
    x = t[:2]
    reveal_type(x)  # tuple[int, int, int], according to this PR?
```

Reply:
Ah, I guess one needs the `TypeVarTuple` stuff. The issue is that for instance `torch.Size` subclasses `tuple`, but adding two `torch.Size` instances gives a `torch.Size` and not a `tuple`.

Reply:
I guess in some sense the issue is really the special casing here, we have two things: `tuple` with a finite set of types and `tuple` with arbitrary size. The `Self` only applies directly to the arbitrary size case. Is there any good way to do a case-distinction for this?

Reply:
Yeah, the issue is partly the extent to which `tuple` is heavily special-cased, partly the fact we can't use `TypeVarTuple` yet in typeshed, and partly the fact that we'd probably also need higher-kinded types to truly express in the stubs what you're going for here :-)

I don't think there's any way in today's type system to express what you're trying to express here, unfortunately :-(

Pytorch could maybe do something like this, as a workaround:

```py
class Size:
    if TYPE_CHECKING:
        def __getitem__(self, __key: slice) -> Self: ...
```

Sorry that I can't offer anything better :/

Reply:
Even for the base class `Sequence` it seems logical that we should have `def __getitem__(self, __s: slice) -> Self`. Tuple with a finite set of types is really quite different to this semantically.

Reply:
Just spitballing here, but it seems like there should be `FiniteSequence(Ts)` and `tuple[T1, ..., Tn]` should be a subtype of `FiniteSequence` whereas `tuple[T, ...]` should be a subtype of `Sequence[T]`.

Reply:
> Even for the base class `Sequence` it seems logical that we should have `def __getitem__(self, __s: slice) -> Self`.

See my comments here on that one: https://github.com/python/typeshed/issues/10727#issuecomment-1723396017